source/common/upstream/upstream_impl.h

```diff
@@ -470,12 +466,23 @@ class ClusterImplBase : public Cluster, protected Logger::Loggable<Logger::Id::u
   virtual void startPreInit() PURE;
 
   /**
-   * Called by every concrete cluster when pre-init is complete. At this point, shared init takes
-   * over and determines if there is an initial health check pass needed, etc.
+   * Called by every concrete cluster when pre-init is complete. At this point,
+   * shared init starts sds_init_manager_ initialization and determines if there
+   * is an initial health check pass needed, etc. onPreInitComplete() may be called
+   * multiple times, but sds_init_manager_ only can be called once.
    */
   void onPreInitComplete();
 
+  /**
+   * Called by every concrete cluster after all Sds api targets registered at SDS init manager are
+   * initialized. At this point, shared init takes over and determines if there is an initial health
+   * check pass needed, etc.
+   */
+  void onSdsInitDone();
+
   Runtime::Loader& runtime_;
+  Server::InitManagerImplPtr
+      sds_init_manager_; // This init manager is only used by SDS API targets.
```

Comment:
Remove comment and just call this `init_manager_`. I'm sure we will use this for other things in the future. (Rename other functions to be generic also)